File: Objc-Webkit-JS/Objc-Webkit-JS/WebKitVC/WeakScriptDelegate.m
//
//  WeakScriptDelegate.m
//  Objc-Webkit-JS
//
//  Created by 小星星 on 2018/8/17.
//  Copyright © 2018年 yangxin. All rights reserved.
//

#import "WeakScriptDelegate.h"


/**
 为了处理内存泄漏，将JS交互代理出来
 */
@implementation WeakScriptDelegate

- (instancetype)initWith: (id<WKScriptMessageHandler>)scriptDelegate
{
    self = [super init];
    if (self) {
        _scriptDelegate = scriptDelegate;
    }
    return self;
}

- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message
{
    [self.scriptDelegate userContentController:userContentController didReceiveScriptMessage:message];
}

@end

File: Swift_And_Js/Swift_And_Js/WeakScriptDelegate.swift
//
//  WeakScriptMessageDelegate.swift
//  WKWebViewJSForSwift
//
//  Created by 小星星 on 2018/8/16.
//  Copyright © 2018年 yangxin. All rights reserved.
//

import UIKit
import WebKit

/// 为了解决内存泄漏，将 ScriptMessageDelegate 作为交互的代理对象，替代web所在的VC
class WeakScriptDelegate: NSObject, WKScriptMessageHandler {
    
    weak var scriptDelegate: WKScriptMessageHandler?
    
    init(_ scriptDelegate: WKScriptMessageHandler) {
        self.scriptDelegate = scriptDelegate
        super.init()
    }
    
    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        scriptDelegate?.userContentController(userContentController, didReceive: message)
    }
    
    deinit {
        print("WeakScriptDelegate is deinit")
    }
}


